git-repository/src/object/mod.rs

```diff
@@ -0,0 +1,149 @@
+        self.try_into()
+    }
+}
+
+pub mod find {
+    use crate::odb;
+
+    pub type Error = odb::compound::find::Error;
+    pub mod existing {
+        use crate::odb;
+
+        pub type Error = odb::pack::find::existing::Error<odb::compound::find::Error>;
+    }
+}
+
+impl<'repo, A> ObjectRef<'repo, A> {
+    pub fn to_owned(&self) -> Object {
+        Object {
+            id: self.id,
+            kind: self.kind,
+            data: self.data.to_owned(),
+        }
+    }
+
+    pub fn into_owned(self) -> Object {
+        Object {
+            id: self.id,
+            kind: self.kind,
+            data: self.data.to_owned(),
+        }
+    }
+
+    pub fn detach(self) -> Object {
+        self.into()
+    }
+}
+
+impl<'repo, A> ObjectRef<'repo, A>
+where
+    A: Access + Sized,
+{
+    pub fn to_commit_iter(&self) -> Option<immutable::CommitIter<'_>> {
+        odb::data::Object::new(self.kind, &self.data).into_commit_iter()
+    }
+
+    pub fn to_tag_iter(&self) -> Option<immutable::TagIter<'_>> {
+        odb::data::Object::new(self.kind, &self.data).into_tag_iter()
+    }
+}
+
+pub mod peel_to_kind {
+    use crate::{
+        object::{peel_to_kind, Kind},
+        Access, ObjectRef,
+    };
+
+    impl<'repo, A> ObjectRef<'repo, A>
+    where
+        A: Access + Sized,
+    {
+        // TODO: tests
+        pub fn peel_to_kind(mut self, kind: Kind) -> Result<Self, peel_to_kind::Error> {
+            loop {
+                match self.kind {
+                    any_kind if kind == any_kind => {
+                        return Ok(self);
+                    }
+                    Kind::Commit => {
+                        let tree_id = self.to_commit_iter().expect("commit").tree_id().expect("valid commit");
+                        let access = self.access;
+                        drop(self);
+                        self = crate::ext::access::object::find_existing_object(access, tree_id)?;
+                    }
+                    Kind::Tag => {
+                        let target_id = self.to_tag_iter().expect("tag").target_id().expect("valid tag");
+                        let access = self.access;
+                        drop(self);
+                        self = crate::ext::access::object::find_existing_object(access, target_id)?;
+                    }
+                    Kind::Tree | Kind::Blob => {
+                        return Err(peel_to_kind::Error::NotFound {
+                            actual: self.kind,
+                            expected: kind,
+                        })
+                    }
+                }
+            }
+        }
+    }
+
+    mod error {
+        use quick_error::quick_error;
+
+        use crate::{object, object::find};
+
+        quick_error! {
+            #[derive(Debug)]
+            pub enum Error {
+                FindExisting(err: find::existing::Error) {
+                    display("A non existing object was encountered during object peeling")
```

Comment:
Tangent, because I came across `find_existing` before: this error means that the repository is probably corrupted, right? Like, we found a reference to the oid, but the actual object doesn't exist. Should this then scream very loudly?

Reply:
I think with sparse checkouts everything is possible, and blobs can usually be missing. Would anybody point a tag to a blob? Probably not. Could they? Probably.
I think libraries shouldn't scream in panics, if this means loudly :), and so far in plumbing crates I found myself at least degenerate the error in the case that objects that should be there aren't to keep error handling easier.

With `git_repository::Easy` panicking on incorrect usage, it's definitely worth thinking about what that means for servers for example and generally, who may panic and when.

Reply:
Ah right, haven't thought about sparse checkouts. I was just thinking about the message text, not panicking. I might suggest one once this actually happens to me :)

Reply:
Yes, please do!
Sparse-Checkouts are a killer feature that I got to have, or else `gitoxide` won't ever be the fastest thing around ever growing mono-repos. Recently git learned to checkout repos with multiple threads, a planned USP now relegated to maintaining parity.